File: module/lunest/Process.lua
---@class lunest.Process
---@field private input lunest.File
---@field private output lunest.File
---@field private input_callbacks table<string, function[]>
local M = {}
---@private
M.__index = M

local json = require("json")

local File = require("lunest.File")

---@param input lunest.File
---@param output lunest.File
---@return self
function M.new(input, output)
    local self = setmetatable({}, M)
    self.input = input
    self.output = output
    self.input_callbacks = {
        Initialize = {},
        Run = {},
        Execute = {},
        Finish = {},
    }
    self:on_finish(function()
        self:close()
    end)
    return self
end

---@param input string
---@param output string
---@return self
function M.open(input, output)
    return M.new(File.open(input, "r"), File.open(output, "a"))
end

function M:close()
    self.input:close()
    self.output:close()
end

function M:loop()
    local buf = ""
    local all_inputs_read = true
    while not self.input:is_closed() do
        local line = self.input:readln()
        if line then
            self:log("line: %s", line)
            all_inputs_read = false
            if line:sub(#line) == "\n" then
                ---@type lunest.Input
                local input = json.decode(buf .. line)
                self:log("calling callbacks of %q", input.t)
                for _, callback in ipairs(self.input_callbacks[input.t] or {}) do
                    callback(input.c)
                end
                buf = ""
            else
                buf = buf .. line
            end
        elseif not all_inputs_read and buf == "" then
            all_inputs_read = true
            self:write({ t = "AllInputsRead" })
        end
    end
end

---@param output lunest.Output
function M:write(output)
    return self.output:writeln(json.encode(output))
end

---@param f fun(input: lunest.Input.Initialize)
function M:on_initialize(f)
    table.insert(self.input_callbacks.Initialize, f)
end

---@param f fun(input: lunest.Input.Run)
function M:on_run(f)
    table.insert(self.input_callbacks.Run, f)
end

---@param f fun(script: string)
function M:on_execute(f)
    table.insert(self.input_callbacks.Execute, f)
end

---@param f fun()
function M:on_finish(f)
    table.insert(self.input_callbacks.Finish, f)
end

---@param s string
---@param ... any
function M:log(s, ...)
    if not self.output:is_closed() then
        return self:write({
            t = "Log",
            c = s:format(...),
        })
    end
end

---@param id string
---@param title string[]
function M:send_test_info(id, title)
    return self:write({
        t = "TestInfo",
        c = {
            id = id,
            title = title,
        },
    })
end

---@param title string[]
function M:notify_test_started(title)
    return self:write({
        t = "TestStarted",
        c = { title = title },
    })
end

---@param title string[]
---@param err lunest.TestError?
function M:notify_test_finished(title, err)
    return self:write({
        t = "TestFinished",
        c = {
            title = title,
            error = err,
        },
    })
end

--- enum
---@alias lunest.Input
---| { t: "Initialize", c: lunest.Input.Initialize }
---| { t: "Run", c: lunest.Input.Run }
---| { t: "Execute", c: string }
---| { t: "Finish", c: nil }
--- enum content
---@class lunest.Input.Initialize
---@field root_dir string
---@field target_files { name: string, path: string }[]
---@field term_width integer
--- enum content
---@class lunest.Input.Run
---@field test_id_filter string[]?
---@field test_mode lunest.TestMode
--- enum
---@alias lunest.TestMode
---| "Run"
---| "SendInfo"

--- enum
---@alias lunest.Output
---| { t: "TestInfo", c: lunest.Output.TestInfo }
---| { t: "TestStarted", c: lunest.Output.TestStarted }
---| { t: "TestFinished", c: lunest.Output.TestFinished }
---| { t: "AllInputsRead", c: nil }
---| { t: "Log", c: string }
--- enum content
---@class lunest.Output.TestInfo
---@field id string
---@field title string[]
--- enum content
---@class lunest.Output.TestStarted
---@field title string[]
--- enum content
---@class lunest.Output.TestFinished
---@field title string[]
---@field error lunest.TestError?

--- struct
---@class lunest.TestError
---@field message string
---@field traceback string
---@field info lunest.TestErrorInfo?

--- enum
---@class lunest.TestErrorInfo
---@field Diff? { left: string, right: string }

return M

File: cli/src/process.rs
use std::fmt;

use crossterm::{style::Stylize, terminal};
use serde::{Deserialize, Serialize};
use tokio::io::AsyncWriteExt;

pub struct Process {
    inner: Option<std::process::Child>,
    input: tokio::fs::File,
    output: crate::buffer::AsyncLineReader<tokio::fs::File>,
}

#[derive(thiserror::Error, Debug)]
pub enum Error {
    #[error(transparent)]
    Io(#[from] std::io::Error),
    #[error("{}", get_exit_error_message(.0))]
    Exit(Option<i32>),
}

impl Process {
    pub async fn spawn(
        cx: &crate::global::Context,
        profile: &crate::profile::Profile,
    ) -> Result<Self, std::io::Error> {
        log::trace!("spawning new process");

        let temp_dir = cx.create_process_dir()?;
        let input_path = temp_dir.join("in.jsonl");
        let output_path = temp_dir.join("out.jsonl");

        let mut cmd = profile.lua_command().clone();
        cmd.program(cx.get_lua_program(cmd.get_program())?)
            .arg(cx.get_main_script())
            .env("LUNEST_IN", &input_path)
            .env("LUNEST_OUT", &output_path)
            .current_dir(cx.root_dir());
        log::debug!("lua command: {}", cmd.display().env(true));

        let child = loop {
            match cmd
                .build()
                .stdout(std::process::Stdio::piped())
                .stderr(std::process::Stdio::piped())
                .spawn()
            {
                Ok(child) => {
                    break child;
                }
                Err(e) if e.kind() == std::io::ErrorKind::ResourceBusy => {
                    log::warn!("failed to spawn the command: {e}");
                    tokio::time::sleep(std::time::Duration::from_millis(100)).await;
                    log::info!("retrying to spawn...");
                }
                Err(e) => {
                    log::error!("failed to spawn the command: {e}");
                    return Err(e);
                }
            }
        };
        log::info!("process spawned as {}", child.id());

        Ok(Self {
            inner: Some(child),
            input: tokio::fs::File::options()
                .create_new(true)
                .append(true)
                .open(input_path)
                .await?,
            output: {
                tokio::fs::File::create(&output_path).await?;
                tokio::fs::File::open(output_path).await?.into()
            },
        })
    }

    pub async fn read(&mut self) -> Result<Option<Output>, std::io::Error> {
        let output = match self.output.read_line().await? {
            crate::buffer::Line::Ok(s) => {
                let out = serde_json::from_str(&s).expect("failed to deserialize an output");
                match &out {
                    Output::Log(s) => log::info!("[log] {s}"),
                    _ => log::debug!("output read: {out:?}"),
                }
                Some(out)
            }
            crate::buffer::Line::NoLF | crate::buffer::Line::Empty => None,
        };
        Ok(output)
    }

    pub async fn write(&mut self, input: &Input) -> Result<(), std::io::Error> {
        log::debug!("writing input: {input:?}");
        let mut json = serde_json::to_vec(input).expect("failed to serialize an input");
        json.extend(b"\n");
        self.input.write_all(&json).await?;
        Ok(())
    }

    pub fn is_running(&mut self) -> Result<bool, Error> {
        let Some(inner) = &mut self.inner else {
            return Ok(true);
        };
        if inner.try_wait()?.is_none() {
            return Ok(true);
        }
        let inner = self.inner.take().unwrap();
        log::info!("process {} already exited", inner.id());

        let out = inner.wait_with_output()?;
        log::debug!("stdout: {}", String::from_utf8_lossy(&out.stdout));
        log::debug!("stderr: {}", String::from_utf8_lossy(&out.stderr));

        match out.status.code() {
            Some(0) => Ok(false),
            code => Err(Error::Exit(code)),
        }
    }
}

impl Drop for Process {
    fn drop(&mut self) {
        if let Some(inner) = &mut self.inner {
            _ = inner.kill();
        }
    }
}

fn get_exit_error_message(code: &Option<i32>) -> String {
    match code {
        Some(n) => format!("spawned process exited with status code {n}"),
        None => "spawned process terminated by signal".into(),
    }
}

#[derive(Debug, Serialize)]
#[serde(tag = "t", content = "c")]
pub enum Input {
    Initialize {
        target_files: Vec<TargetFile>,
        root_dir: std::path::PathBuf,
        term_width: u16,
    },
    Run {
        test_id_filter: Option<Vec<String>>,
        test_mode: TestMode,
    },
    Execute(std::path::PathBuf),
    Finish,
}

#[derive(Debug, Serialize)]
pub struct TargetFile {
    path: std::path::PathBuf,
    name: String,
}

#[derive(Debug, Serialize)]
pub enum TestMode {
    Run,
    SendInfo,
}

impl TargetFile {
    pub fn from_path(path: std::path::PathBuf, root_dir: &std::path::Path) -> Self {
        let rel = path.strip_prefix(root_dir).unwrap_or(&path);
        let name = rel.display().to_string().replace('\\', "/");
        Self { path, name }
    }
}

#[derive(Debug, Deserialize)]
#[serde(tag = "t", content = "c")]
pub enum Output {
    TestInfo(TestInfo),
    TestStarted(TestStarted),
    TestFinished(TestFinished),
    AllInputsRead,
    Log(String),
}

fn fmt_title(title: &[String]) -> String {
    title.join(&" :: ".grey().to_string())
}

#[derive(Debug, Deserialize)]
pub struct TestInfo {
    pub id: String,
    pub title: Vec<String>,
}

impl fmt::Display for TestInfo {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{}", fmt_title(&self.title))
    }
}

#[derive(Debug, Deserialize)]
pub struct TestStarted {
    title: Vec<String>,
}

impl fmt::Display for TestStarted {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(
            f,
            "{}{} {}",
            fmt_title(&self.title),
            ":".grey(),
            "RUNNING".cyan().bold()
        )
    }
}

#[derive(Debug, Deserialize)]
pub struct TestFinished {
    title: Vec<String>,
    error: Option<TestError>,
}

impl TestFinished {
    pub fn success(&self) -> bool {
        self.error.is_none()
    }
}

impl fmt::Display for TestFinished {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{}", terminal::Clear(terminal::ClearType::UntilNewLine))?;
        write!(f, "{}{} ", fmt_title(&self.title), ":".grey())?;
        if let Some(err) = &self.error {
            write!(f, "{}\n{}", "ERR".red().bold(), err)
        } else {
            write!(f, "{}", "OK".green().bold())
        }
    }
}

#[derive(Debug, Deserialize)]
pub struct TestError {
    message: String,
    traceback: String,
    info: Option<TestErrorInfo>,
}

#[derive(Debug, Deserialize)]
pub enum TestErrorInfo {
    Diff { left: String, right: String },
}

impl fmt::Display for TestError {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        writeln!(f, "{}\n", self.message.as_str().bold())?;
        if let Some(info) = &self.info {
            writeln!(f, "{info}")?;
        }
        writeln!(f, "{}:", "  stack traceback".bold())?;
        writeln!(f, "{}", self.traceback)?;
        Ok(())
    }
}

impl fmt::Display for TestErrorInfo {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match &self {
            TestErrorInfo::Diff { left, right } => {
                writeln!(
                    f,
                    "{} ({} {} {}):",
                    "  difference".bold(),
                    "-left".red(),
                    "/".grey(),
                    "+right".green()
                )?;
                let delete = "-".red();
                let insert = "+".green();
                let diff = similar::TextDiff::from_lines(left, right);
                for change in diff.iter_all_changes() {
                    let line = change.value();
                    use similar::ChangeTag::*;
                    match change.tag() {
                        Equal => write!(f, " {}", line.grey())?,
                        Delete => write!(f, "{delete}{}", line.red())?,
                        Insert => write!(f, "{insert}{}", line.green())?,
                    }
                    if !line.ends_with('\n') {
                        writeln!(f)?;
                    }
                }
            }
        }
        Ok(())
    }
}
